src/relay/backend/contrib/dnnl/query_layout.cc

```diff
@@ -0,0 +1,340 @@
+                                        std::string out_shape, std::string paddings,
+                                        std::string strides, std::string dilates, std::string G) {
+  dnnl::engine eng(dnnl::engine::kind::cpu, 0);
+  dnnl::stream s(eng);
+  using tag = dnnl::memory::format_tag;
+  using dt = dnnl::memory::data_type;
+
+  dnnl::memory::dim groups = std::stoi(G);
+  dnnl::memory::dims weight_dims_ = str2dims(weight_shape, input_size);
+  dnnl::memory::dims weight_dims = weight_dims_;
+  if (groups > 1) {
+    if (weight_dims_.size() == 5) {
+      weight_dims = {weight_dims_[0] * weight_dims_[1], weight_dims_[2], weight_dims_[3],
+                     weight_dims_[4]};
+    } else {
+      weight_dims[1] = weight_dims[1] * groups;
+    }
+  }
+  dnnl::memory::dims out_dims = str2dims(out_shape, input_size);
+  dnnl::memory::dims padding_dims = str2dims(paddings, 2 * (input_size - 2));
+  dnnl::memory::dims padding_dims_l(padding_dims.begin(),
+                                    padding_dims.begin() + padding_dims.size() / 2);
+  dnnl::memory::dims padding_dims_r(padding_dims.end() - padding_dims.size() / 2,
+                                    padding_dims.end());
+  dnnl::memory::dims strides_dims = str2dims(strides, input_size - 2);
+  dnnl::memory::dims dilates_dims = str2dims(dilates, input_size - 2);
+
+  dnnl::memory::dims input_dims = out_dims;
+  input_dims[1] = weight_dims[1];
+  for (int i = 2; i < input_size; i++) {
+    dnnl::memory::dim K = weight_dims[i];
+    dnnl::memory::dim S = strides_dims[i - 2];
+    dnnl::memory::dim D = dilates_dims[i - 2] - 1;
+    dnnl::memory::dim PL = padding_dims_l[i - 2];
+    dnnl::memory::dim PR = padding_dims_r[i - 2];
+    dnnl::memory::dim DK = 1 + (K - 1) * (D + 1);
+    dilates_dims[i - 2] = D;
+    input_dims[i] = out_dims[i] * S - PL - PR + DK - 1;
+  }
+
+  dnnl::memory::dims conv_src_dims = input_dims;
+  dnnl::memory::dims conv_weights_dims = weight_dims;
+  if (groups > 1) {
+    conv_weights_dims = {groups, out_dims[1] / groups, input_dims[1] / groups};
+    conv_weights_dims.insert(conv_weights_dims.end(), weight_dims.begin() + 2, weight_dims.end());
+  }
+  dnnl::memory::dims conv_dst_dims = out_dims;
+  dnnl::memory::dims conv_strides = strides_dims;
+  dnnl::memory::dims conv_dilates = dilates_dims;
+  dnnl::memory::dims conv_padding_l = padding_dims_l;
+  dnnl::memory::dims conv_padding_r = padding_dims_r;
+
+  auto conv_src_md = dnnl::memory::desc({conv_src_dims}, dt::f32, tag::any);
+  auto conv_weights_md = dnnl::memory::desc({conv_weights_dims}, dt::f32, tag::any);
+  auto conv_dst_md = dnnl::memory::desc({conv_dst_dims}, dt::f32, tag::any);
+
+  auto conv_desc = dnnl::convolution_forward::desc(
+      dnnl::prop_kind::forward_inference, dnnl::algorithm::convolution_direct, conv_src_md,
+      conv_weights_md, conv_dst_md, conv_strides, conv_dilates, conv_padding_l, conv_padding_r);

```

Comment:
Using `tag::any` for src and weights here, but using the layout from JSON at dnnl_json_runtime, is this the cause of the "OI"-"IO" problem?

Reply:
Any progress on the "OI"-"IO" problem?

Reply:
1. This issue can be concluded as, the queried layout for group conv is `debcA16a`. However, we cannot attain layout format `tag::debcA16a`, but can only attain `tag::decbA16a` in `dnnl.hpp`. Note that this issue only appears when onednnv2.5 is used.
2. I have checked onednnv2.5 benchdnn, the preferred layout is `debcA16a` for all group convs. I think this is the OneDNN issue.
3. I can fix this issue by setting the `weight_layout` as `tag::any` in `dnnl_json_runtime.cc` to get the correct layout. But consider the community is using onednnv2.2, I suggest not to add the solution in this PR.

Reply:
dnnl format `debcA16a` = tvm layout `HWOIG16g`;
dnnl format `decbA16a` = tvm layout `HWIOG16g`